scripts/prune_lm_dir.py

```diff
@@ -328,6 +351,160 @@ def FinalizeOutput(final_work_out):
+    Returns:
+        A tuple with the first element is a bool variable indicating whether we already got the right threshold.
+        And the remaining are updated value for (step, iter, recovery_step, recovery_size)
+    Example Usage:
+
+        step = 0
+        recovery_step = step
+        recovery_size = 0
+        iter = 0
+
+        threshold = init_threshold()
+        while True:
+            threshold = EstimateThreshold(args.target_num_ngrams)
+
+            (success, step, iter, recovery_step, recovery_size) = \
+                    IterateOnce(threshold, step, iter, recovery_step, recovery_size)
+            if success:
+                break
+
+    """
+    global logprob_changes, steps, final_num_ngrams
+
+    if step > 0:
+        scale = threshold / args.threshold
+        steps += 'prune*{0} EM EM EM'.format(scale).split()
+
+    # Prune step
+    logprob_changes.append(RunStep(step, in_step=recovery_step))
+    step += 1
+    while step < len(steps): # EM steps
+      logprob_changes.append(RunStep(step))
+      step += 1
+    iter += 1
+
+    if MatchTargetSize(final_num_ngrams):
+        return (True, step, iter, recovery_step, recovery_size)
+
+    if iter > args.max_iter:
+        sys.exit("prune_lm_dir.py: Too many iterations, please set a higher threshold and rerun.")
+
+    # always prune from LM larger than target_num_ngrams
+    if final_num_ngrams > args.target_num_ngrams and (recovery_size <= 0 or final_num_ngrams < recovery_size):
+        recovery_step = step
+        recovery_size = final_num_ngrams
+
+    return (False, step, iter, recovery_step, recovery_size)
+
+class LinearEstimator(object):
+    """Estimate the coeffients of a line using two most recent points
+
+    Example Usage:
+
+        ls = LinearEstimator()
+        ls.AddPoint(x0, y0)
+        ls.AddPoint(x1, y1)
+
+        while True:
+            xn = Update()
+            yn = ls.Estimate(xn)
+
+            if Match(yn):
+                break
+
+            ls.AddPoint(xn, yn)
+    """
+    def __init__(self):
+        self.switch = True
+        self.x0 = 0.0
+        self.x1 = 0.0
+        self.y0 = 0.0
+        self.y1 = 0.0
+        self.n = 0
+
+    def AddPoint(self, x, y):
+        if self.switch:
+            self.x0 = x
+            self.y0 = y
+        else:
+            self.x1 = x
+            self.y1 = y
+        self.switch = not self.switch
+        self.n += 1
+
+    def Estimate(self, x):
+        assert self.n >= 2
+        alpha = (self.y1 - self.y0) / (self.x1 - self.x0)
+        beta = self.y0 - alpha * self.x0
+
+        return alpha * x + beta
+
+# find threshold in order to match the target-num-ngrams with final LM
+#
+# Here we fit a linear of log(num-ngrams) versus log(threshold)
+# with the latest two points, and approach the target-num-ngrams gradually.
+def FindThreshold():
+    global final_num_ngrams
+
+    step = 0
+    recovery_step = step
```

Comment:
please document recovery_step at this point, e.g.
   # recovery_step is the index of the most recent step of pruning that resulted in
   # a model with num-ngrams > target_size.

and
   # recovery_size will be set to the number of ngrams in the model at step 'recovery_step'